vector/src/main/java/im/vector/app/features/login/qr/QrCodeLoginViewModel.kt

```diff
@@ -58,20 +71,60 @@ class QrCodeLoginViewModel @AssistedInject constructor(
     }
 
     private fun handleOnQrCodeScanned(action: QrCodeLoginAction.OnQrCodeScanned) {
-        if (isValidQrCode(action.qrCode)) {
-            setState {
-                copy(
-                        connectionStatus = QrCodeLoginConnectionStatus.ConnectingToDevice
-                )
+        Timber.tag(TAG).d("Scanned code: ${action.qrCode}")
+
+        val rendezvous = try { Rendezvous.buildChannelFromCode(action.qrCode) } catch (t: Throwable) {
+            Timber.tag(TAG).e(t, "Error occurred during sign in")
+            if (t is RendezvousError) {
+                onFailed(t.reason)
+            } else {
+                onFailed(RendezvousFailureReason.Unknown)
+            }
+            return
+        }
+
+        setState {
+            copy(
+                    connectionStatus = QrCodeLoginConnectionStatus.ConnectingToDevice
+            )
+        }
+
+        _viewEvents.post(QrCodeLoginViewEvents.NavigateToStatusScreen)
+
+        viewModelScope.launch(Dispatchers.IO) {
+            try {
+                val confirmationCode = rendezvous.startAfterScanningCode()
+                Timber.tag(TAG).i("Established secure channel with checksum: $confirmationCode")
+                confirmationCode?.let {
+                    onConnectionEstablished(it)
+                    val session = rendezvous.waitForLoginOnNewDevice(authenticationService)
+                    onSigningIn()
+                    session?.let {
```

Comment:
Should we do something when `session` is `null`?

Reply:
Following a refactor this is actually never `null`. I've therefore set it as non-null and removed the lets.